### PR TITLE
NEM Support

### PR DESCRIPTION
**Pull request by saleemrashid**
Add NEM transaction serialization and NEM encryption. All functionality has complete unit tests.

### CODE REVIEW

File: bip32.c [hunk 1]
```diff
@@ -94,7 +99,7 @@ int hdnode_from_xprv(uint32_t depth, uint32_t child_num, const uint8_t *chain_co
 	const curve_info *info = get_curve_by_name(curve);
 	if (info == 0) {
 		failed = true;
-	} else {
+	} else if (info->params) {
```
Comment (jhoenicke) [inline bip32.c]: This fix should probably go to master.  Can you put it in a separate pull request?

Currently this function is only called with "secp256k1", so it is not be exploitable, right?

Comment (saleemrashid) [inline bip32.c]: It's only used with Ed25519 in the unit tests so it's not exploitable. But I will make a new pull request for it.

**Comment by prusnak** [comment]
Please rebase after https://github.com/trezor/trezor-crypto/pull/104 has been merged.

**Comment by prusnak** [comment]
@saleemrashid are you OK with me squashing the changes into one commit?

Also I am not sure whether we should have `nem.c` and `nem.h` in this repository. It has quite a lot of application logic and we have this in trezor-mcu for other coins.

**Comment by saleemrashid** [comment]
@prusnak If you want to squash the changes, I would say a few commits (e.g. BIP32, BIP32 tests, serialization and serialization tests).

Serialization tests especially are quite large so would be good to keep them in their own commit. 

Unlike the code in trezor-mcu, this is very portable and is written more as a general purpose library for NEM (does not have any global state, has bounds checking, is endian agnostic, etc.) so it makes sense to have it here.

Also, while we can use device tests, it is good having unit tests that can be run under e.g. Valgrind, which isn't trivial if the code is moved to trezor-mcu.

**Comment by saleemrashid** [comment]
> It has quite a lot of application logic

Could you give some specific examples as I was explicitly trying to avoid this (e.g. you don't need a private key for serialization, you can serialize encrypted transactions with just the ciphertext). 
The only thing that you could say that about would probably be the multisignature serialization (because it does the hashing, rather than accepting a hash), but it would be not too difficult to change this, if you want.

File: nem.c [hunk 2]
```diff
@@ -0,0 +1,590 @@
...
+}
+
+static inline bool nem_write_mosaic_bool(nem_transaction_ctx *ctx, const char *name, bool value) {
+	return nem_write_mosaic_str(ctx, name, value ? "true" : "false");
+}
+
+static inline bool nem_write_mosaic_u64(nem_transaction_ctx *ctx, const char *name, uint64_t value) {
+	char buffer[21];
+
+	char *s = &buffer[sizeof(buffer) - 1];
```
Comment (saleemrashid) [inline nem.c]: I wrote this before the `bn_format` redactor, but now we can use `bn_format_uint64` here

Comment (prusnak) [reply nem.c]: Please push another commit which refactors this to use `bn_format_uint64`, thanks!

Comment (saleemrashid) [reply nem.c]: @prusnak Fixed in 065731a

**Comment by saleemrashid** [comment]
@prusnak I'm going to squash the changes now and force push over this branch

**Comment by saleemrashid** [comment]
@prusnak Rebased and squashed

**Comment by prusnak** [comment]
Looks nice, thanks, merged!

**Pull request merged**